### PR TITLE
Update MNIST example w.r.t. KFP 2.5.0

### PR DESCRIPTION
**Pull request by roopak-rastogi**
Provide a clear and concise description of the content changes you're proposing. List all the changes you are making
to the content.

- Updated section ___ with ___.
- Added new subsection ___ under ___.
- Removed outdated information from ___.
- ...

Closes #...

> If there is no issue related to this PR, kindly create one first to describe the motivation behind these changes.

Checklist:

- [x] I have checked that my enhancements are not duplicates of existing content changes or additions.
- [x] I have tested the changes in a working environment to ensure they function as intended.
- [ ] I have followed the [style guide](https://github.com/HPEEzmeral/ezua-tutorials/blob/develop/CONTRIBUTING.md#style-guide)
      outlined in the contribution guidelines.

Reviewer's Tasks (for maintainers reviewing this PR):

- [ ] Verify that the tutorial functions correctly in a live environment.
- [ ] Verify that the updated content aligns with the [style guide](https://github.com/HPEEzmeral/ezua-tutorials/blob/develop/CONTRIBUTING.md#style-guide)
      in the contribution guidelines.
- [ ] Check for consistency, grammar, and clarity throughout the updated content.
- [ ] Check that the related GitHub issue is up-to-date.

### CODE REVIEW

File: Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb [hunk 1]
```diff
@@ -41,7 +41,8 @@
     "# %env https_proxy=http://<proxy_host>:<port>\n",
     "# Install required packages (Kubeflow Pipelines and Katib SDK).\n",
     "import sys\n",
-    "!{sys.executable} -m pip install \"pyarrow>7.0.0\" \"kfp>=1.8.0,<=1.8.22\" kubeflow-katib==0.16.0"
+    "# !{sys.executable} -m pip install \"pyarrow>7.0.0\" \"kfp>=1.8.0,<=1.8.22\" kubeflow-katib==0.16.0\n",
```
Comment (CybervisionVolodymyrPogorelov) [inline Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb:44]: I think that we don't need to install kfp in the example itself. It should be available in the notebook by default.

**Review by CybervisionVolodymyrPogorelov — CHANGES_REQUESTED** [review]
We need to use KFP that goes in the notebook by default

File: Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb (hunk 1 above)
Comment (roopak-rastogi) [reply Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb:44]: Agreed. I'll remove it.

Comment (roopak-rastogi) [reply Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb]: Done

**Comment by roopak-rastogi** [comment]
> 1. Copy-pasting is bad - either figure ou,t how to get volume name from volume operation object, or define a constant.
> 2. Passing parameters that are unused is bad, but maybe you will use that volume operation object.
> 3. Commenting code make sense only, if it is expected that user will uncomment it for any reason (like setting proxy), but old code should just be deleted.
> 4. Using of the notebooks image is a clever idea, but it lacks airgap prefix. @AyushSinha5588 has added it to notebook as env var in past. Also that image is too big for what you are trying to achieve here. I've just solved similar problem for an another use case - https://jira-pro.it.hpe.com:8443/browse/EZAF-5507. We might want to add a new image for that or reuse some of kfp images like kfp-launcher or kfp-diver.

Addressed all the remarks. 
For the point 2: Passing parameters that are unused is bad, but maybe you will use that volume operation object.
Currently, model_volume is being passed as an argument to the functions, which is not being utilized/referenced directly. Due to the open issue of accessing passed arguments value (as of now, accessing passed arguments value is not accessed, rather than it shows `{{channel:task=;name=name;type=String;}}` ), once the related issue is resolved in the future release, the value can be referenced directly from the arguments passed. Therefore, leaving the argument as it is.

File: Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb (hunk 1 above)
Comment (umka1332) [reply Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb]: It actually was worse then we don't need to - it was installation of the previous version, so if user uncomment that, it could brake whole example.

p.s. it was also already covered by my point from the comment above.:

> Commenting code make sense only, if it is expected that user will uncomment it for any reason (like setting proxy), but old code should just be deleted.

File: Data-Science/Kubeflow/MNIST-Digits-Recognition/component/kubeflow-launcher-component.yaml [hunk 2]
```diff
@@ -9,9 +9,9 @@ inputs:
...
 - {name: CleanPodPolicy,              type: String,     default: Running,   description: 'Defines the policy for cleaning up pods after the TFJob completes.'}
 - {name: PS Spec,                     type: JsonObject, default: '{}',      description: 'TFJob ps replicaSpecs.'}
-- {name: Worker Spec,                 type: JsonObject, default: '{}',      description: 'TFJob worker replicaSpecs.'}
-- {name: Chief Spec,                  type: JsonObject, default: '{}',      description: 'TFJob chief replicaSpecs.'}
-- {name: Evaluator Spec,              type: JsonObject, default: '{}',      description: 'TFJob evaluator replicaSpecs.'}
+- {name: Worker Spec,                 type: String, default: '{}',      description: 'TFJob worker replicaSpecs.'}
+- {name: Chief Spec,                  type: String, default: '{}',      description: 'TFJob chief replicaSpecs.'}
+- {name: Evaluator Spec,              type: String, default: '{}',      description: 'TFJob evaluator replicaSpecs.'}
 - {name: Tfjob Timeout Minutes,       type: Integer,    default: 1440,      description: 'Time in minutes to wait for the TFJob to complete.'}
```
Comment (umka1332) [inline Data-Science/Kubeflow/MNIST-Digits-Recognition/component/kubeflow-launcher-component.yaml:15]: Please fix indentation - just like it is done for Tfjob Timeout Minutes

Comment (roopak-rastogi) [reply Data-Science/Kubeflow/MNIST-Digits-Recognition/component/kubeflow-launcher-component.yaml]: Done.

File: Data-Science/Kubeflow/MNIST-Digits-Recognition/component/kubeflow-launcher-component.yaml [hunk 3]
```diff
@@ -9,9 +9,9 @@ inputs:
 - {name: ttl Seconds After Finished,  type: Integer,    default: -1,        description: 'Defines the TTL for cleaning up finished TFJobs.'}
 - {name: CleanPodPolicy,              type: String,     default: Running,   description: 'Defines the policy for cleaning up pods after the TFJob completes.'}
 - {name: PS Spec,                     type: JsonObject, default: '{}',      description: 'TFJob ps replicaSpecs.'}
-- {name: Worker Spec,                 type: JsonObject, default: '{}',      description: 'TFJob worker replicaSpecs.'}
-- {name: Chief Spec,                  type: JsonObject, default: '{}',      description: 'TFJob chief replicaSpecs.'}
-- {name: Evaluator Spec,              type: JsonObject, default: '{}',      description: 'TFJob evaluator replicaSpecs.'}
+- {name: Worker Spec,                 type: String,     default: '{}',      description: 'TFJob worker replicaSpecs.'}
+- {name: Chief Spec,                  type: String,     default: '{}',      description: 'TFJob chief replicaSpecs.'}
+- {name: Evaluator Spec,              type: String,     default: '{}',      description: 'TFJob evaluator replicaSpecs.'}
```
Comment (umka1332) [inline Data-Science/Kubeflow/MNIST-Digits-Recognition/component/kubeflow-launcher-component.yaml:14]: `PS Spec` should probably be also changed.

Comment (umka1332) [reply Data-Science/Kubeflow/MNIST-Digits-Recognition/component/kubeflow-launcher-component.yaml:14]: We are not using it, but the same is true for `Evaluator Spec`

File: Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb [hunk 4]
```diff
@@ -52,7 +52,8 @@
    "outputs": [],
    "source": [
     "# Specify the name of the user mounted directory\n",
-    "user_mounted_dir_name = \"user\""
+    "user_mounted_dir_name = \"user\"\n",
+    "airgap_registry=\"lr1-bd-harbor-registry.mip.storage.hpecorp.net/develop/\""
```
Comment (umka1332) [inline Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb:56]: We need to get it from env var

File: Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb [hunk 5]
```diff
@@ -568,19 +575,18 @@
...
     "    \n",
     "    # Create volume to train and serve the model.\n",
-    "    model_volume_op = dsl.VolumeOp(\n",
-    "        name=volume_str,\n",
-    "        resource_name=volume_str,\n",
+    "    model_volume_op = kfp.kubernetes.CreatePVC(\n",
+    "        pvc_name=volume_str,\n",
     "        size=\"1Gi\",\n",
-    "        modes=dsl.VOLUME_MODE_RWO\n",
+    "        storage_class_name=\"<storage_class_name>\",\n",
```
Comment (umka1332) [inline Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb:588]: We can use something like this instead of expecting user to know this and notice to change.
```
storage_class_name = !kubectl get pvc user-pvc -o jsonpath={.spec.storageClassName}
storage_class_name = storage_class_name[0]
```

File: Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb [hunk 6]
```diff
@@ -39,9 +39,11 @@
    "source": [
     "# Uncomment & specify proxy value if applicable\n",
     "# %env https_proxy=http://<proxy_host>:<port>\n",
+    "# Comma separated list of IP address\n",
+    "# %env no_proxy=\"<host1>:<port1>,<host2>:<port2>,...\"\" \n",
```
Comment (umka1332) [inline Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb:43]: there could be not only IPs, but also IPs with masks and text hostnames. And afaik, it can not contain ports contrrary to httss_proxy.

File: Data-Science/Kubeflow/MNIST-Digits-Recognition/component/kubeflow-launcher-component.yaml (hunk 3 above)
Comment (roopak-rastogi) [reply Data-Science/Kubeflow/MNIST-Digits-Recognition/component/kubeflow-launcher-component.yaml]: Done

File: Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb (hunk 4 above)
Comment (roopak-rastogi) [reply Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb]: Added the airgap_registry from env in the base_image path

File: Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb (hunk 5 above)
Comment (roopak-rastogi) [reply Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb]: Updated the code with the above code snippet.

File: Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb (hunk 6 above)
Comment (roopak-rastogi) [reply Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb]: Updated the comment and also added list in the `no_proxy`  as an example to provide better clarity. If anything else is also needed, please update me.

File: Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb [hunk 7]
```diff
@@ -39,9 +39,11 @@
    "source": [
     "# Uncomment & specify proxy value if applicable\n",
     "# %env https_proxy=http://<proxy_host>:<port>\n",
+    "# After default.svc append comma separated list of IP address, hostname, CIDR block, service name, etc.\n",
+    "# %env no_proxy=\"localhost,.cluster.local,.svc,.default.svc, <host>:<port>, x.x.x.x/y,...\"\n",
     "# Install required packages (Kubeflow Pipelines and Katib SDK).\n",
     "import sys\n",
-    "!{sys.executable} -m pip install \"pyarrow>7.0.0\" \"kfp>=1.8.0,<=1.8.22\" kubeflow-katib==0.16.0"
+    "!{sys.executable} -m pip install kfp[kubernetes]"
```
Comment (akravacyber) [inline Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb:46]: Do we need to install it actually? If we do not install `kfp[kubernetes]`, then example will fail, right? Looks strange, I've thought that we install the whole kfp package. If example is not working without it, I'm ok to leave it here. But in future in notebooks we can install this in all notebook images I think

File: Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb [hunk 8]
```diff
@@ -568,19 +576,19 @@
...
+    "    storage_class_name = !kubectl get pvc user-pvc -o jsonpath={.spec.storageClassName}\n",
     "    # Create volume to train and serve the model.\n",
-    "    model_volume_op = dsl.VolumeOp(\n",
-    "        name=volume_str,\n",
-    "        resource_name=volume_str,\n",
+    "    model_volume_op = kfp.kubernetes.CreatePVC(\n",
+    "        pvc_name=volume_str,\n",
     "        size=\"1Gi\",\n",
-    "        modes=dsl.VOLUME_MODE_RWO\n",
+    "        storage_class_name=storage_class_name[0],\n",
```
Comment (akravacyber) [inline Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb:590]: Do we need to pass a storage class name? I think it is not required, and if we omit it here - default storage class will be used by kubernetes

**Review by akravacyber — COMMENTED** [review]
Actually looks good, left some minor comments

File: Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb (hunk 7 above)
Comment (roopak-rastogi) [reply Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb:46]: We can add it in the notebooks in future. Have tested it without installing kfp[kubernetes] it was failing as we are calling it for creating the PVC.

File: Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb (hunk 8 above)
Comment (roopak-rastogi) [reply Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb:590]: Without storage_class_name parameter, PVC was getting created in `standard` storage_class, since this storage_class is not present with this name, therefore, PVC and the TF jobs were going in Pending state. Same was seen when we tried with empty ("") storage_class_name.

File: Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb (hunk 7 above)
Comment (akravacyber) [reply Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb:46]: @roopak-rastogi thank you, got it, okay. It is fine

File: Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb (hunk 8 above)
Comment (akravacyber) [reply Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb:590]: Hmm, got it, okay, Than yes, we can leave it as is

File: Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb (hunk 7 above)
Comment (umka1332) [reply Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb]: Now, due to issue with `storageClassName` we are creating pvc with `kubectl`, so `kfp.kubernetes` module is not needed any more.

File: Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb [hunk 9]
```diff
@@ -372,10 +376,16 @@
    "outputs": [],
    "source": [
     "# This function converts Katib Experiment HP results to args.\n",
-    "def convert_katib_results(katib_results) -> str:\n",
+    "\n",
+    "from kfp import components\n",
+    "@dsl.component(\n",
+    "    install_kfp_package = False,\n",
+    "    base_image = f\"{os.environ['AIRGAP_REGISTRY']}gcr.io/mapr-252711/kubeflow/python-kfp:3.7-2.7.0\",\n",
```
Comment (umka1332) [inline Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb:383]: Specifying base image and disabling kfp install are not needed any more - we have merged PR, that allows package installation:

- https://github.hpe.com/hpe/ezaf/pull/2855
- https://github.com/mapr/private-manifests/pull/584
- and most impotent - https://github.com/mapr/private-kubeflow-pipelines/pull/32

**Review by ask664 — APPROVED** [review]

File: Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb [hunk 10]
```diff
@@ -388,12 +396,23 @@
...
     "# You should define the TFJob name, namespace, number of training steps, output of Katib and model volume tasks in the arguments.\n",
-    "def create_tfjob_task(tfjob_name, tfjob_namespace, training_steps, katib_op, model_volume_op):\n",
+    "def create_tfjob_task(tfjob_name, tfjob_namespace, training_steps, katib_op, model_volume_op, http_proxy, no_proxy):\n",
     "    import json\n",
     "    # Get parameters from the Katib Experiment.\n",
     "    # Parameters are in the format \"--tf-learning-rate=0.01 --tf-batch-size=100\"\n",
-    "    convert_katib_results_op = components.func_to_container_op(convert_katib_results)\n",
-    "    best_hp_op = convert_katib_results_op(katib_op.output)\n",
+    "    best_hp_op = convert_katib_results(katib_results = katib_op.output)\n",
+    "    # Setting the proxy and no_proxy to allow download of the required packages inside the component. Uncomment the below code, if proxies are specified above.\n",
```
Comment (umka1332) [inline Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb:404]: Please remove or update the comment. In case of update please notice that it is not related to `set_caching_options` that sits just below it.

File: Data-Science/Kubeflow/MNIST-Digits-Recognition/pvc.yaml [hunk 11]
```diff
@@ -0,0 +1,10 @@
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: pvc-name
+spec:
+  accessModes:
+   - ReadWriteOnce
+  resources:
+   requests:
+    storage: 1Gi
```
Comment (umka1332) [inline Data-Science/Kubeflow/MNIST-Digits-Recognition/pvc.yaml:10]: Please add a newline at the end of file.

File: Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb (hunk 10 above)
Comment (roopak-rastogi) [reply Data-Science/Kubeflow/MNIST-Digits-Recognition/mnist_katib_tf_kserve_example.ipynb]: Updated the comment and moved the `set_caching_options` part before the comment.

File: Data-Science/Kubeflow/MNIST-Digits-Recognition/pvc.yaml (hunk 11 above)
Comment (roopak-rastogi) [reply Data-Science/Kubeflow/MNIST-Digits-Recognition/pvc.yaml]: Done

File: Data-Science/Kubeflow/MNIST-Digits-Recognition/component/kubeflow-launcher-component.yaml [hunk 12]
```diff
@@ -8,10 +8,10 @@ inputs:
...
 - {name: CleanPodPolicy,              type: String,     default: Running,   description: 'Defines the policy for cleaning up pods after the TFJob completes.'}
-- {name: PS Spec,                     type: JsonObject, default: '{}',      description: 'TFJob ps replicaSpecs.'}
-- {name: Worker Spec,                 type: JsonObject, default: '{}',      description: 'TFJob worker replicaSpecs.'}
-- {name: Chief Spec,                  type: JsonObject, default: '{}',      description: 'TFJob chief replicaSpecs.'}
-- {name: Evaluator Spec,              type: JsonObject, default: '{}',      description: 'TFJob evaluator replicaSpecs.'}
+- {name: PS Spec,                     type: String,     default: '{}',      description: 'TFJob ps replicaSpecs.'}
+- {name: Worker Spec,                 type: String,     default: '{}',      description: 'TFJob worker replicaSpecs.'}
+- {name: Chief Spec,                  type: String,     default: '{}',      description: 'TFJob chief replicaSpecs.'}
+- {name: Evaluator Spec,              type: String,     default: '{}',      description: 'TFJob evaluator replicaSpecs.'}
```
[the remark this answers is not in the document]
Comment (roopak-rastogi) [reply Data-Science/Kubeflow/MNIST-Digits-Recognition/component/kubeflow-launcher-component.yaml:15]: Updated the file with the mentioned comment.

**Review by umka1332 — APPROVED** [review]

**Review by CybervisionVolodymyrPogorelov — APPROVED** [review]
